File: leetcode/ReverseWordsInString/revword.cpp
#include<iostream>
#include<string>

using namespace std;

class Solution
{

        void reverseString(string &s, int lo, int hi)
        {
            char temp;

            while(lo < hi)
            {
                temp = s[lo];
                s[lo] = s[hi];
                s[hi] = temp;
                lo++;
                hi--;
            }
        }

        void removeMulSpaces(string &s, int lo, int hi)
        {
            string new_s;
            
            for(int i = lo; i <= hi; i++)
            {
                if(isspace(s[i]))
                    continue;
                else
                {
                    int l = i;
                    int h = l;
                    
                    for(; h <= hi; h++)
                    {
                        if(!isspace(s[h]))
                            new_s.push_back(s[h]);
                        else
                        {
                            new_s.push_back(' ');
                            break;
                        }
                    }
                    
                    i = h; //update i
                }
            }
            
            s.swap(new_s);
        }
    
    public:
    
        void reverseWords(string &s)
        {

            // reverse entire string
            
            reverseString(s, 0 , s.length() - 1);        
             
            // reverse individual words 
            
            for(int i = 0; i < (int)s.length(); i++)
            {
               if( isspace(s[i]) )
                   continue;
               
               else
               {
                    int lo = i;
                    int hi = lo;

                    for(; hi < (int)s.length(); hi++)
                    {
                        if( !isspace(s[hi]) )
                            continue;
                        else
                        {   
                            break;
                        }
                    }
                    
                    i = hi; // update i

                    reverseString(s, lo, hi - 1);
               }
            }
        
            // remove trailing, leading and multiple spaces .. as per question specification */
            
            removeMulSpaces(s, 0, s.length() - 1);
        }
};


int main()
{
    string s("");

    Solution soln;

    soln.reverseWords(s);

    cout << s << endl;
    return 0;
}

File: leetcode/ReverseWordsInString/revword.py
#https://oj.leetcode.com/problems/reverse-words-in-a-string/
#status = AC
import sys

class Solution(object):
    # @param s, a string
    # @return a string

    def reverseWords(self,s):

        # reverse entire string

        s = s[::-1]

        ls = s.split()
        new_ls = []
        
        # reverse individual words

        for x in ls:
            new_ls.append(x[::-1])
        
        new_s = ""

        for x in new_ls:
            new_s = new_s + x + " "
    
        #delete last char as it is a space 

        return new_s[:-1]


soln = Solution()

s = "  the    sky   is   blue   "
print soln.reverseWords(s)

